### PR TITLE
Allow TLS configuration update wherever possible.

### PR DESCRIPTION
**Pull request by tanmayja**
Allow TLS update in certain scenarios like:
1 - ca-file to ca-path change in existing and in-use TLS config.
2- Adding, removing and modifying TLS config, which is not being used in any network conf.

### CODE REVIEW

File: api/v1/aerospikecluster_validating_webhook.go [hunk 1]
```diff
@@ -1396,6 +1392,47 @@ func validateAerospikeConfigUpdate(
...
+		}
+
+		for _, tls := range oldTLS.([]interface{}) {
+			tlsMap := tls.(map[string]interface{})
+
+			oldCAFile, oldCAFileOK := tlsMap["ca-file"]
+			if oldCAFileOK {
+				oldTLSCAFileMap[tlsMap["name"].(string)] = oldCAFile.(string)
+			}
+		}
```
Comment (abhishekdwivedi3060) [inline api/v1/aerospikecluster_validating_webhook.go:1420]: I think this oldTLSCAFileMap should be populated for only those TLS which were used earlier in any sub-context. If TLS is just added and not used then we can allow any type of update on that

File: api/v1/aerospikecluster_validating_webhook.go [hunk 2]
```diff
@@ -1396,6 +1392,47 @@ func validateAerospikeConfigUpdate(
...
+
+		for _, tls := range newTLS.([]interface{}) {
+			tlsMap := tls.(map[string]interface{})
+			if caFile, ok := oldTLSCAFileMap[tlsMap["name"].(string)]; ok {
```
Comment (abhishekdwivedi3060) [inline api/v1/aerospikecluster_validating_webhook.go:1424]: Check `usedTLS.Has(tlsMap["name"].(string))` before if condition and continue if not used

File: api/v1/aerospikecluster_validating_webhook.go [hunk 3]
```diff
@@ -1396,6 +1392,47 @@ func validateAerospikeConfigUpdate(
...
+				newCAFile, newCAFileOK := tlsMap["ca-file"]
+				if newCAFileOK && newCAFile.(string) != caFile && usedTLS.Has(tlsMap["name"].(string)) {
+					return fmt.Errorf("cannot change ca-file of used tls")
+				}
```
Comment (abhishekdwivedi3060) [inline api/v1/aerospikecluster_validating_webhook.go:1428]: Check if new TLS has `ca-path` set or not when `newCAFileOK` is false

**Review by abhishekdwivedi3060 — CHANGES_REQUESTED** [review]

File: test/cluster_test.go [hunk 4]
```diff
@@ -331,7 +336,176 @@ func DeployClusterWithSyslog(ctx goctx.Context) {
...
+					Expect(err).ToNot(HaveOccurred())
+
+					network := aeroCluster.Spec.AerospikeConfig.Value["network"].(map[string]interface{})
+					tlsList := network["tls"].([]interface{})
+					usedTLS := tlsList[0].(map[string]interface{})
+					usedTLS["name"] = "aerospike-a-0.test-runner2"
+					tlsList[0] = usedTLS
+					network["tls"] = tlsList
+					aeroCluster.Spec.AerospikeConfig.Value["network"] = network
+					err = updateCluster(k8sClient, ctx, aeroCluster)
```
Comment (abhishekdwivedi3060) [inline test/cluster_test.go:484]: We should call `updateClusterWithTO` with small timeout to reduce test time as we are sure test-case is going to fail. Change wherever applicable

File: test/cluster_test.go [hunk 5]
```diff
@@ -331,7 +336,176 @@ func DeployClusterWithSyslog(ctx goctx.Context) {
...
+					network = aeroCluster.Spec.AerospikeConfig.Value["network"].(map[string]interface{})
+					tlsList = network["tls"].([]interface{})
+					usedTLS = tlsList[0].(map[string]interface{})
+					usedTLS["ca-file"] = "/etc/aerospike/secret/fb_cert.pem"
+					tlsList[0] = usedTLS
+					network["tls"] = tlsList
+					aeroCluster.Spec.AerospikeConfig.Value["network"] = network
+					err = updateCluster(k8sClient, ctx, aeroCluster)
+					Expect(err).Should(HaveOccurred())
+				},
```
Comment (abhishekdwivedi3060) [inline test/cluster_test.go:502]: Should we add steps to check when both ca-file and ca-path are updated in cluster?

File: api/v1/aerospikecluster_validating_webhook.go [hunk 6]
```diff
@@ -1396,6 +1390,82 @@ func validateAerospikeConfigUpdate(
...
+
+			_, newCAPathOK := tlsMap["ca-path"]
+			newCAFile, newCAFileOK := tlsMap["ca-file"]
+
+			oldCAFile, oldCAFileOK := oldTLSCAFileMap[tlsMap["name"].(string)]
+			_, oldCAPathOK := oldTLSCAPathMap[tlsMap["name"].(string)]
+
+			if (oldCAFileOK || oldCAPathOK) && !(newCAPathOK || newCAFileOK) {
+				return fmt.Errorf(
+					"cannot remove `ca-file` or `ca-path` from tls",
```
Comment (abhishekdwivedi3060) [inline api/v1/aerospikecluster_validating_webhook.go:1454]: nit: 
```suggestion
					"cannot remove used `ca-file` or `ca-path` from tls",
```

File: api/v1/aerospikecluster_validating_webhook.go [hunk 7]
```diff
@@ -1396,6 +1390,82 @@ func validateAerospikeConfigUpdate(
...
+				)
+			}
+
+			if oldCAFileOK {
+				if newCAFileOK && newCAFile.(string) != oldCAFile {
```
Comment (abhishekdwivedi3060) [inline api/v1/aerospikecluster_validating_webhook.go:1459]: Can be clubbed into one

**Review by abhishekdwivedi3060 — CHANGES_REQUESTED** [review]

File: test/cluster_test.go (hunk 4 above)
Comment (tanmayja) [reply test/cluster_test.go:484]: When webhook fails(which is this case), updateCluster is not gonna take much time.

**Review by abhishekdwivedi3060 — APPROVED** [review]

**Review by sud82 — APPROVED** [review]

**Pull request merged**